// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package cirq.google.engine.client.quantum_v1alpha1.proto;

import "google/api/annotations.proto";
import "google/cloud/quantum_v1alpha1/proto/quantum.proto";
import "google/protobuf/duration.proto";
import "google/protobuf/empty.proto";
import "google/protobuf/field_mask.proto";
import "google/api/client.proto";

// -
service QuantumEngineService {
  option (.google.api.default_host) = "quantum.googleapis.com";
  option (.google.api.oauth_scopes) = "https://www.googleapis.com/auth/cloud-platform";

  // -
  rpc CreateQuantumProgram(CreateQuantumProgramRequest) returns (QuantumProgram) {
    option (.google.api.http) = {
      post: "/v1alpha1/{parent=projects/*}/programs"
      body: "quantum_program"
    };
  }

  // -
  rpc GetQuantumProgram(GetQuantumProgramRequest) returns (QuantumProgram) {
    option (.google.api.http) = {
      get: "/v1alpha1/{name=projects/*/programs/*}"
    };
  }

  // -
  rpc ListQuantumPrograms(ListQuantumProgramsRequest) returns (ListQuantumProgramsResponse) {
    option (.google.api.http) = {
      get: "/v1alpha1/{parent=projects/*}/programs"
    };
  }

  // -
  rpc DeleteQuantumProgram(DeleteQuantumProgramRequest) returns (.google.protobuf.Empty) {
    option (.google.api.http) = {
      delete: "/v1alpha1/{name=projects/*/programs/*}"
    };
  }

  // -
  rpc UpdateQuantumProgram(UpdateQuantumProgramRequest) returns (QuantumProgram) {
    option (.google.api.http) = {
      patch: "/v1alpha1/{name=projects/*/programs/*}"
      body: "quantum_program"
    };
  }

  // -
  rpc CreateQuantumJob(CreateQuantumJobRequest) returns (QuantumJob) {
    option (.google.api.http) = {
      post: "/v1alpha1/{parent=projects/*/programs/*}/jobs"
      body: "quantum_job"
    };
  }

  // -
  rpc GetQuantumJob(GetQuantumJobRequest) returns (QuantumJob) {
    option (.google.api.http) = {
      get: "/v1alpha1/{name=projects/*/programs/*/jobs/*}"
    };
  }

  // -
  rpc ListQuantumJobs(ListQuantumJobsRequest) returns (ListQuantumJobsResponse) {
    option (.google.api.http) = {
      get: "/v1alpha1/{parent=projects/*/programs/*}/jobs"
    };
  }

  // -
  rpc DeleteQuantumJob(DeleteQuantumJobRequest) returns (.google.protobuf.Empty) {
    option (.google.api.http) = {
      delete: "/v1alpha1/{name=projects/*/programs/*/jobs/*}"
    };
  }

  // -
  rpc UpdateQuantumJob(UpdateQuantumJobRequest) returns (QuantumJob) {
    option (.google.api.http) = {
      patch: "/v1alpha1/{name=projects/*/programs/*/jobs/*}"
      body: "quantum_job"
    };
  }

  // -
  rpc CancelQuantumJob(CancelQuantumJobRequest) returns (.google.protobuf.Empty) {
    option (.google.api.http) = {
      post: "/v1alpha1/{name=projects/*/programs/*/jobs/*}:cancel"
      body: "*"
    };
  }

  // -
  rpc ListQuantumJobEvents(ListQuantumJobEventsRequest) returns (ListQuantumJobEventsResponse) {
    option (.google.api.http) = {
      get: "/v1alpha1/{parent=projects/*/programs/*/jobs/*}/events"
    };
  }

  // -
  rpc GetQuantumResult(GetQuantumResultRequest) returns (QuantumResult) {
    option (.google.api.http) = {
      get: "/v1alpha1/{parent=projects/*/programs/*/jobs/*}/result"
    };
  }

  // -
  rpc ListQuantumProcessors(ListQuantumProcessorsRequest) returns (ListQuantumProcessorsResponse) {
    option (.google.api.http) = {
      get: "/v1alpha1/{parent=projects/*}/processors"
    };
  }

  // -
  rpc GetQuantumProcessor(GetQuantumProcessorRequest) returns (QuantumProcessor) {
    option (.google.api.http) = {
      get: "/v1alpha1/{name=projects/*/processors/*}"
    };
  }

  // -
  rpc ListQuantumCalibrations(ListQuantumCalibrationsRequest) returns (ListQuantumCalibrationsResponse) {
    option (.google.api.http) = {
      get: "/v1alpha1/{parent=projects/*/processors/*}/calibrations"
    };
  }

  // -
  rpc GetQuantumCalibration(GetQuantumCalibrationRequest) returns (QuantumCalibration) {
    option (.google.api.http) = {
      get: "/v1alpha1/{name=projects/*/processors/*/calibrations/*}"
    };
  }

  // -
  rpc CreateQuantumReservation(CreateQuantumReservationRequest) returns (QuantumReservation) {
    option (.google.api.http) = {
      post: "/v1alpha1/{parent=projects/*/processors/*}/reservations"
      body: "quantum_reservation"
    };
  }

  // -
  rpc CancelQuantumReservation(CancelQuantumReservationRequest) returns (QuantumReservation) {
    option (.google.api.http) = {
      post: "/v1alpha1/{name=projects/*/processors/*/reservations/*}:cancel"
      body: "*"
    };
  }

  // -
  rpc DeleteQuantumReservation(DeleteQuantumReservationRequest) returns (.google.protobuf.Empty) {
    option (.google.api.http) = {
      delete: "/v1alpha1/{name=projects/*/processors/*/reservations/*}"
    };
  }

  // -
  rpc GetQuantumReservation(GetQuantumReservationRequest) returns (QuantumReservation) {
    option (.google.api.http) = {
      get: "/v1alpha1/{name=projects/*/processors/*/reservations/*}"
    };
  }

  // -
  rpc ListQuantumReservations(ListQuantumReservationsRequest) returns (ListQuantumReservationsResponse) {
    option (.google.api.http) = {
      get: "/v1alpha1/{parent=projects/*/processors/*}/reservations"
    };
  }

  // -
  rpc UpdateQuantumReservation(UpdateQuantumReservationRequest) returns (QuantumReservation) {
    option (.google.api.http) = {
      patch: "/v1alpha1/{name=projects/*/processors/*/reservations/*}"
      body: "quantum_reservation"
    };
  }

  // -
  rpc QuantumRunStream(stream QuantumRunStreamRequest) returns (stream QuantumRunStreamResponse) {
  }

  // -
  rpc ListQuantumReservationGrants(ListQuantumReservationGrantsRequest) returns (ListQuantumReservationGrantsResponse) {
    option (.google.api.http) = {
      get: "/v1alpha1/{parent=projects/*}/reservationGrant"
    };
  }

  // -
  rpc ReallocateQuantumReservationGrant(ReallocateQuantumReservationGrantRequest) returns (QuantumReservationGrant) {
    option (.google.api.http) = {
      post: "/v1alpha1/{name=projects/*/reservationGrant/*}:reallocate"
      body: "*"
    };
  }

  // -
  rpc ListQuantumReservationBudgets(ListQuantumReservationBudgetsRequest) returns (ListQuantumReservationBudgetsResponse) {
    option (.google.api.http) = {
      get: "/v1alpha1/{parent=projects/*}/reservationBudgets"
    };
  }

  // -
  rpc ListQuantumTimeSlots(ListQuantumTimeSlotsRequest) returns (ListQuantumTimeSlotsResponse) {
    option (.google.api.http) = {
      get: "/v1alpha1/{parent=projects/*/processors/*}/timeSlots"
    };
  }
}

// -
message CreateQuantumJobRequest {
  // -
  string parent = 1;

  // -
  QuantumJob quantum_job = 2;

  // -
  bool overwrite_existing_run_context = 3;
}

// -
message GetQuantumJobRequest {
  // -
  string name = 1;

  // -
  bool return_run_context = 2;
}

// -
message ListQuantumJobsRequest {
  // -
  string parent = 1;

  // -
  int32 page_size = 2;

  // -
  string page_token = 3;

  // -
  string filter = 4;
}

// -
message ListQuantumJobsResponse {
  // -
  repeated QuantumJob jobs = 1;

  // -
  string next_page_token = 2;
}

// -
message DeleteQuantumJobRequest {
  // -
  string name = 1;
}

// -
message UpdateQuantumJobRequest {
  // -
  string name = 1;

  // -
  QuantumJob quantum_job = 2;

  // -
  .google.protobuf.FieldMask update_mask = 3;
}

// -
message CancelQuantumJobRequest {
  // -
  string name = 1;
}

// -
message ListQuantumJobEventsRequest {
  // -
  string parent = 1;

  // -
  int32 page_size = 2;

  // -
  string page_token = 3;
}

// -
message ListQuantumJobEventsResponse {
  // -
  repeated QuantumJobEvent events = 1;

  // -
  string next_page_token = 2;
}

// -
message GetQuantumResultRequest {
  // -
  string parent = 1;
}

// -
message CreateQuantumProgramRequest {
  // -
  string parent = 1;

  // -
  QuantumProgram quantum_program = 2;

  // -
  bool overwrite_existing_source_code = 3;
}

// -
message GetQuantumProgramRequest {
  // -
  string name = 1;

  // -
  bool return_code = 2;
}

// -
message ListQuantumProgramsRequest {
  // -
  string parent = 1;

  // -
  int32 page_size = 2;

  // -
  string page_token = 3;

  // -
  string filter = 4;
}

// -
message ListQuantumProgramsResponse {
  // -
  repeated QuantumProgram programs = 1;

  // -
  string next_page_token = 2;
}

// -
message DeleteQuantumProgramRequest {
  // -
  string name = 1;

  // -
  bool delete_jobs = 2;
}

// -
message UpdateQuantumProgramRequest {
  // -
  string name = 1;

  // -
  QuantumProgram quantum_program = 2;

  // -
  .google.protobuf.FieldMask update_mask = 3;
}

// -
message ListQuantumProcessorsRequest {
  // -
  string parent = 1;

  // -
  int32 page_size = 2;

  // -
  string page_token = 3;

  // -
  string filter = 4;
}

// -
message ListQuantumProcessorsResponse {
  // -
  repeated QuantumProcessor processors = 1;

  // -
  string next_page_token = 2;
}

// -
message GetQuantumProcessorRequest {
  // -
  string name = 1;
}

// -
message ListQuantumCalibrationsRequest {
  // -
  string parent = 1;

  // -
  int32 page_size = 2;

  // -
  string page_token = 3;

  // -
  string filter = 4;
}

// -
message ListQuantumCalibrationsResponse {
  // -
  repeated QuantumCalibration calibrations = 1;

  // -
  string next_page_token = 2;
}

// -
message GetQuantumCalibrationRequest {
  // -
  string name = 1;
}

// -
message CreateQuantumReservationRequest {
  // -
  string parent = 1;

  // -
  QuantumReservation quantum_reservation = 2;
}

// -
message CancelQuantumReservationRequest {
  // -
  string name = 1;
}

// -
message DeleteQuantumReservationRequest {
  // -
  string name = 1;
}

// -
message GetQuantumReservationRequest {
  // -
  string name = 1;
}

// -
message ListQuantumReservationsRequest {
  // -
  string parent = 1;

  // -
  int32 page_size = 2;

  // -
  string page_token = 3;

  // -
  string filter = 4;
}

// -
message ListQuantumReservationsResponse {
  // -
  repeated QuantumReservation reservations = 1;

  // -
  string next_page_token = 2;
}

// -
message UpdateQuantumReservationRequest {
  // -
  string name = 1;

  // -
  QuantumReservation quantum_reservation = 2;

  // -
  .google.protobuf.FieldMask update_mask = 3;
}

// -
message QuantumRunStreamRequest {
  // -
  string message_id = 1;

  // -
  string parent = 2;

  // -
  oneof request {
    // -
    CreateQuantumProgramAndJobRequest create_quantum_program_and_job = 3;

    // -
    CreateQuantumJobRequest create_quantum_job = 4;

    // -
    GetQuantumResultRequest get_quantum_result = 5;
  }
}

// -
message CreateQuantumProgramAndJobRequest {
  // -
  string parent = 1;

  // -
  QuantumProgram quantum_program = 2;

  // -
  QuantumJob quantum_job = 3;
}

// -
message QuantumRunStreamResponse {
  // -
  string message_id = 1;

  // -
  oneof response {
    // -
    StreamError error = 2;

    // -
    QuantumJob job = 3;

    // -
    QuantumResult result = 4;
  }
}

// -
message StreamError {
  // -
  enum Code {
    // -
    CODE_UNSPECIFIED = 0;

    // -
    INTERNAL = 1;

    // -
    INVALID_ARGUMENT = 2;

    // -
    PERMISSION_DENIED = 3;

    // -
    PROGRAM_ALREADY_EXISTS = 4;

    // -
    JOB_ALREADY_EXISTS = 5;

    // -
    PROGRAM_DOES_NOT_EXIST = 6;

    // -
    JOB_DOES_NOT_EXIST = 7;

    // -
    PROCESSOR_DOES_NOT_EXIST = 8;

    // -
    INVALID_PROCESSOR_FOR_JOB = 9;

    // -
    RESERVATION_REQUIRED = 10;
  }

  // -
  Code code = 1;

  // -
  string message = 2;
}

// -
message ListQuantumReservationGrantsRequest {
  // -
  string parent = 1;

  // -
  int32 page_size = 2;

  // -
  string page_token = 3;

  // -
  string filter = 4;
}

// -
message ListQuantumReservationGrantsResponse {
  // -
  repeated QuantumReservationGrant reservation_grants = 1;

  // -
  string next_page_token = 2;
}

// -
message ReallocateQuantumReservationGrantRequest {
  // -
  string name = 1;

  // -
  string source_project_id = 2;

  // -
  string target_project_id = 3;

  // -
  .google.protobuf.Duration duration = 4;
}

// -
message ListQuantumReservationBudgetsRequest {
  // -
  string parent = 1;

  // -
  int32 page_size = 2;

  // -
  string page_token = 3;

  // -
  string filter = 4;
}

// -
message ListQuantumReservationBudgetsResponse {
  // -
  repeated QuantumReservationBudget reservation_budgets = 1;

  // -
  string next_page_token = 2;
}

// -
message ListQuantumTimeSlotsRequest {
  // -
  string parent = 1;

  // -
  int32 page_size = 2;

  // -
  string page_token = 3;

  // -
  string filter = 4;
}

// -
message ListQuantumTimeSlotsResponse {
  // -
  repeated QuantumTimeSlot time_slots = 1;

  // -
  string next_page_token = 2;
}
